Sources/NIOCore/NIOTimer.swift

```diff
@@ -0,0 +1,94 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A type that handles timer callbacks scheduled with ``EventLoop/setTimer(for:_:)-5e37g``.
+///
+/// - Seealso: ``EventLoop/setTimer(for:_:)-5e37g``.
+public protocol NIOTimerHandler {
+    func timerFired(loop: any EventLoop)
+}
+
+/// An opaque handle that can be used to cancel a timer.
+///
+/// Users cannot create an instance of this type; it is returned by ``EventLoop/setTimer(for:_:)-5e37g``.
+///
+/// - Seealso: ``EventLoop/setTimer(for:_:)-5e37g``.
+public struct NIOTimer {
+    @usableFromInline
+    enum Backing {
+        /// A task created using `EventLoop.scheduleTask(deadline:_:)`, used by default for `EventLoop` implementations.
+        case scheduledTask(Scheduled<Void>)
+        /// An identifier for a timer, used by `EventLoop` implementations that conform to `CustomTimerImplementation`.
+        case custom(eventLoop: any NIOCustomTimerImplementation, id: UInt64)
+    }
+
+    @usableFromInline
+    var backing: Backing
+
+    fileprivate init(_ scheduled: Scheduled<Void>) {
+        self.backing = .scheduledTask(scheduled)
+    }
+
+    @inlinable
+    init(_ eventLoop: any NIOCustomTimerImplementation, id: UInt64) {
+        self.backing = .custom(eventLoop: eventLoop, id: id)
+    }
+
+    /// Cancel the timer associated with this handle.
+    @inlinable
+    public func cancel() {
```

Comment:
@glbrntt Thinking we could make this one safe to call from anywhere cheaply, since it doesn't return a value, we won't have to deal with the ELF? We can just make `NIOCustomTimerImplementation.cancelTimer(_:)` do event loop hop, if necessary?